[GLib] Support for loading websettings from a config file

philn opened the pull request:
#### 4d653b68af47298e0454fb62fb15a52f56c184ae
<pre>
[GLib] Support for loading websettings from a config file
<a href="https://bugs.webkit.org/show_bug.cgi?id=273043">https://bugs.webkit.org/show_bug.cgi?id=273043</a>

Reviewed by Michael Catanzaro, Carlos Garcia Campos and Adrian Perez de Castro.

New WebKitSettings API allowing to apply settings declared in a .ini style file. Based on a similar
feature implemented in Cog by Adrian Perez de Castro &lt;aperez@igalia.com&gt;.

* Source/WTF/wtf/glib/GUniquePtr.h:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webkit_settings_apply_from_key_file):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Tools/MiniBrowser/gtk/main.c:
(main):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitSettingsApplyFromConfigFile):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/278688@main">https://commits.webkit.org/278688@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/657710df3447dc651ac446ee95df3519adabe1bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51313 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3655 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54578 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2003 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53616 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36931 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1682 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41795 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53412 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28262 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44241 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22913 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25587 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1499 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47556 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1572 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56166 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26428 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1469 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49190 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27671 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44304 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48358 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/11220 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28561 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27406 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Tools/MiniBrowser/gtk/main.c:
```diff
@@ -76,6 +76,7 @@ static gboolean enableITP;
 static gboolean exitAfterLoad;
 static gboolean webProcessCrashed;
 static gboolean printVersion;
+static char* configFile;
```
So this file is a little confusing due to the mix of different code styles. With regards to the asterisks, your pull request uses a mix of our C++ style (where asterisk hangs on the type) vs. our C style (where asterisk hangs on the variable name, like GNOME style). The rest of this file puts it on the variable name.

```suggestion
static char *configFile;
```

philn replied:
Oops, I thought the style-checker was reporting these :)

mcatanzaro commented on Tools/MiniBrowser/gtk/main.c:
```diff
@@ -972,6 +974,77 @@ static void activate(GApplication *application, WebKitSettings *webkitSettings)
     gtk_widget_show(GTK_WIDGET(mainWindow));
 }
 
+gboolean webkit_settings_apply_from_key_file(WebKitSettings *settings, GKeyFile *key_file, const char *group, GError ** error)
```
```suggestion
gboolean webkit_settings_apply_from_key_file(WebKitSettings *settings, GKeyFile *key_file, const char *group, GError **error)
```

carlosgcampos replied:
Maybe it makes sense to add this to the api? load/save methods to GKeyFile/path

philn replied:
I'm not sure about save, but yes load at least would be a good start.

mcatanzaro commented on Tools/MiniBrowser/gtk/main.c:
```diff
@@ -972,6 +974,77 @@ static void activate(GApplication *application, WebKitSettings *webkitSettings)
+            }
+
+            case G_TYPE_STRING: {
+                g_autofree char* value = g_key_file_get_string(key_file, group, properties[i]->name, &lookup_error);
```
```suggestion
                g_autofree char *value = g_key_file_get_string(key_file, group, properties[i]->name, &lookup_error);
```

mcatanzaro commented on Tools/MiniBrowser/gtk/main.c:
```diff
@@ -972,6 +974,77 @@ static void activate(GApplication *application, WebKitSettings *webkitSettings)
+    return TRUE;
+}
+
+static gboolean load_settings(WebKitSettings *settings, GKeyFile *key_file, GError ** error)
```
```suggestion
static gboolean load_settings(WebKitSettings *settings, GKeyFile *key_file, GError **error)
```

mcatanzaro commented on Tools/MiniBrowser/gtk/main.c:
```diff
@@ -1024,6 +1097,25 @@ int main(int argc, char *argv[])
 
+    if (configFile) {
+        g_autoptr(GFile) file = g_file_new_for_commandline_arg(configFile);
+        g_autofree char* configFilePath = g_file_get_path(file);
```
```suggestion
        g_autofree char *configFilePath = g_file_get_path(file);
```

mcatanzaro commented on Tools/MiniBrowser/gtk/main.c:
```diff
@@ -972,6 +974,77 @@ static void activate(GApplication *application, WebKitSettings *webkitSettings)
+
+    GObjectClass *klass = G_OBJECT_GET_CLASS(settings);
+    unsigned n_properties = 0;
+    g_autofree GParamSpec* const* properties = g_object_class_list_properties(klass, &n_properties);
```
```suggestion
    g_autofree GParamSpec * const *properties = g_object_class_list_properties(klass, &n_properties);
```

philn commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+ *
+ * Returns: %TRUE if the settings were correctly applied or %FALSE otherwise.
+ *
+ * Since: 2.44
```
`2.46`

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+ *
+ * Since: 2.44
+ */
+gboolean webkit_settings_apply_from_key_file(WebKitSettings* settings, GKeyFile* keyFile, GError ** error)
```
Could we not hardcode the name of the section to be `websettings`, and have a function parameter instead? If we want to have a default section name, we can allow the parameter to be nullable.

```suggestion
gboolean webkit_settings_apply_from_key_file(WebKitSettings* settings, GKeyFile* keyFile, const char* sectionName, GError ** error)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+ *
+ * Since: 2.46
+ */
+gboolean webkit_settings_apply_from_key_file(WebKitSettings* settings, GKeyFile* keyFile, const gchar* groupName, GError ** error)
```
```suggestion
gboolean webkit_settings_apply_from_key_file(WebKitSettings* settings, GKeyFile* keyFile, const gchar* groupName, GError** error)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+ */
+gboolean webkit_settings_apply_from_key_file(WebKitSettings* settings, GKeyFile* keyFile, const gchar* groupName, GError ** error)
+{
+    g_return_val_if_fail(WEBKIT_IS_SETTINGS(settings), FALSE);
+    g_return_val_if_fail(keyFile, FALSE);
```
We should check group name too, I guess it can be nullptr. And error, it should be nullptr or point to nullptr

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
 }
+
+/**
+ * webkit_settings_apply_from_key_file:
```
Do we want this to apply settings at any time to an existing settings, or to create a new settings object with its values loaded from a key file?

aperezdc replied:
I think it's more flexible to apply to an existing settings object, that way you can e.g. retrieve the default settings from a webview and update it, instead of always being forced to create a new object. If one wants a fresh settings object, creating one by hand with default and then calling this function is just one line of code more so I don't think it's worth it to opimize for the “create fresh settings loading file” use-case while making the API less flexible.

philn replied:
This API does the former. I suppose the latter could also be a valid use-case but it can be covered using this API already, on application side.

carlosgcampos replied:
Now that I see the use cases of this api, I think it's fine this way. It was not a matter of optimizing it, it's just that I was thinking more of an API to load/save settings to make them permanent.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+ * webkit_settings_apply_from_key_file:
+ * @settings: a #WebKitSettings
+ * @key_file: a #GKeyFile
+ * @group_name: Name of the group to read from @key_file
```
Why limiting to a group? If we have a key file with several groups we need to call this for every group?

philn replied:
Yes. For instance you could have 2 groups in the same file, one for prod, one for debugging and depending on user's choice, either one of the groups would be loaded.

aperezdc replied:
The group may contain any setting, the idea is that you can have an INI-like file to configure your browser, and populate the settings from one of its sections. Like this:

```ini
[options]
homepage = http://webkitgtk.org
download-path = ~/Downloads

[webkit-settings]
enable-developer-extras = true
enable-resizable-text-areas = true
```

By allowing indicating the `web-settings` group we allow settings loaded by `webkit_settings_apply_from_key_file()` to coexist with other program settings, without forcing a particular group name.

For example in the future we could add `webkit_settings_apply_features_from_key_file()` and the same INI file could have a `[webkit-features]` group, as well.

carlosgcampos replied:
I see, both use cases make sense indeed.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+ * @group_name: Name of the group to read from @key_file
+ * @error: return location for error or %NULL to ignore
+ *
+ * Reads the contents of the given @group_name from the given @key_file and apply the value of
+ * each key/value to the corresponding property on the @settings.
```
I'm not sure that's exactly what this does, or I'm not understanding it. We iterate the WebKitSetting properties, not the key file, failing if the key file doesn't have a property, which requires to load all properties?

carlosgcampos replied:
We should also document how can this fail and what happens when it fails. It seems this is not atomic, so if we fail to set a property, the previous ones are set anyway, but we stop so the next one won't even if they are valid.

philn replied:
That algorithm was Adrian's, I didn't change it.

Maybe it'd be simpler to iterate on the key/values instead?

aperezdc replied:
@carlosgcampos If the key file does not have an entry for a property, it's skipped (see comment in line 4171). Properties are iterated to know which key names to check. It could also be done the other way around: iterating over the existing keys and looking up the corresponding property by name.

Regarding atomicity, I suppose if we wanted to make sure all applied settings are valid we are probably going to need two passes, one to validate and type-check the values in the `GKeyFile`, storing them in an arrays of names and `GValue`s, and then applying them all at once using [g_object_setv()](https://docs.gtk.org/gobject/method.Object.setv.html)... Maybe stuff for a follow-up?

carlosgcampos replied:
> @carlosgcampos If the key file does not have an entry for a property, it's skipped (see comment in line 4171). Properties are iterated to know which key names to check. It could also be done the other way around: iterating over the existing keys and looking up the corresponding property by name.

I've seen the comment, but I don't understand how it works. It seems we iterate the properties and fail we fail to find it in the key file. Am I understanding the code wrong?

> Regarding atomicity, I suppose if we wanted to make sure all applied settings are valid we are probably going to need two passes, one to validate and type-check the values in the `GKeyFile`, storing them in an arrays of names and `GValue`s, and then applying them all at once using [g_object_setv()](https://docs.gtk.org/gobject/method.Object.setv.html)... Maybe stuff for a follow-up?

We are not in a hurry to land this, I prefer to land a version that does what we want with tests checking those cases.

aperezdc replied:
I wrote this a good while ago, but trying to think like myself in the past, probably the reasoning was “I will need the `GParamSpec` for the property anyway, so if I iterate over the properties I already get that _and_ the name at once”.

Feel free to change it, but repeatedly calling `g_object_find_property()` will do a linear search for each property.

aperezdc replied:
Oh, I just remembered, there's also this gem inside `gobject.c`, which means strings for keys dynamically allocated by `GKeyFile` will _not_ match any property string name:

```c
/* This uses pointer comparisons with @property_name, so
 * will only work with string literals. */
static inline GParamSpec *
find_pspec (GObjectClass *class,
            const char   *property_name)
{
          // ...
          if (pspecs[i].name == property_name)
            return pspecs[i].pspec;

          // ...
}
```

Notice there is no `strcmp()` call there! 😱

This means we cannot reliable pass strings which have not been interned by GObject as property names to retrieve property information.

aperezdc replied:
> > @carlosgcampos If the key file does not have an entry for a property, it's skipped (see comment in line 4171). Properties are iterated to know which key names to check. It could also be done the other way around: iterating over the existing keys and looking up the corresponding property by name.
> 
> I've seen the comment, but I don't understand how it works. It seems we iterate the properties and fail we fail to find it in the key file. Am I understanding the code wrong?

The algorithm is basically:

```
  for each *property* of the object class:
     if *keyfile* errored while checking if there's a key named after *property*:
        return with error

     if *keyfile* did _not_ contain a key named after *property*:
        skip to the next property

    convert the value for the key to the property type
    if the conversion suceeded:
       set property to the convered value
    else:
       return with error
```

I hope this helps.

> > Regarding atomicity, I suppose if we wanted to make sure all applied settings are valid we are probably going to need two passes, one to validate and type-check the values in the `GKeyFile`, storing them in an arrays of names and `GValue`s, and then applying them all at once using [g_object_setv()](https://docs.gtk.org/gobject/method.Object.setv.html)... Maybe stuff for a follow-up?
> 
> We are not in a hurry to land this, I prefer to land a version that does what we want with tests checking those cases.

Sure.

carlosgcampos replied:
Ah! it makes sense now, thanks for the clarification. It's weird a function returning FALSE and not setting error, but it's a query function.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+                g_propagate_error(error, lookupError.release());
+                return FALSE;
+            }
+            continue; // Setting missing in GKeyFile, skip it.
```
Setting missing in key file is not what makes `g_key_file_has_key` to fail?

aperezdc replied:
Missing keys cause the function to return `FALSE` and leave `error` untouched. As per [the docs](https://docs.gtk.org/glib/method.KeyFile.has_key.html):

> Note that this function does not follow the rules for GError strictly; the return value both carries meaning and signals an error. To use this function, you must pass a GError pointer in error, and check whether it is not NULL to see if an error occurred.

Meaning:

- Result is `TRUE` → no error, key exists.
- Result is `FALSE`:
  - If error is `NULL` → no error, key does not exist.
  - If error is non-`NULL` → error, we don't know if key exists or not.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+            gboolean value = g_key_file_get_boolean(keyFile, groupName, propertyName, &lookupError.outPtr());
+            if (!value && lookupError) {
+                g_propagate_error(error, lookupError.release());
+                return FALSE;
```
If we want to make parse errors fatal, I think we should parse first and apply later, to make sure we either apply all setting or none. Also if we really want to apply settings on existing object, we should freeze notify before applying settings and thaw after.

aperezdc replied:
Good idea about freeze/thaw to avoid signal storms 👌🏼

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
```diff
@@ -562,6 +562,12 @@ webkit_settings_get_experimental_features                      (void);
+webkit_settings_apply_from_key_file                            (WebKitSettings *settings,
+                                                                GKeyFile       *key_file,
+                                                                const gchar    *group_name,
+                                                                GError        **error);
```
```suggestion
WEBKIT_API gboolean
webkit_settings_apply_from_key_file                            (WebKitSettings  *settings,
                                                                GKeyFile        *key_file,
                                                                const gchar     *group_name,
                                                                GError         **error);
```

(One space between the longest type and the leftmost asterisk)

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
```diff
@@ -458,6 +459,70 @@ void testWebKitFeatures(Test* test, gconstpointer)
+void testWebKitSettingsApplyFromConfigFile(Test* test, gconstpointer)
+{
+    GRefPtr<WebKitSettings> settings = adoptGRef(webkit_settings_new());
+    GKeyFile* key_file = g_key_file_new();
```
Add a GUniquePtr specialization for it?

philn replied:
There is one already actually.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+ *
+ * Supported value types are strings, booleans and unsigned integers.
+ *
+ * Returns: %TRUE if the settings were correctly applied or %FALSE otherwise.
```
```suggestion
 * Returns: %TRUE if the settings were correctly applied or %FALSE on error.
```

To avoid any doubt that `FALSE` corresponds to `error` being set.

mcatanzaro reviewed (commented):
I've only reviewed the API. The API looks good.

carlosgcampos commented on Source/WTF/wtf/glib/GUniquePtr.h:
```diff
@@ -47,6 +47,7 @@ using GUniquePtr = std::unique_ptr<T, U>;
     macro(GList, g_list_free) \
     macro(GSList, g_slist_free) \
     macro(GPatternSpec, g_pattern_spec_free) \
+    macro(GParamSpec, g_free) \
```
I don't think we need this, g_free is the default for GUniquePtr

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+    unsigned totalProperties = 0;
+    GUniquePtr<GParamSpec*> properties(g_object_class_list_properties(klass, &totalProperties));
+
+    if (!g_key_file_has_group(keyFile, groupName)) {
+        g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_INVALID_VALUE, "The key file has no %s group", groupName);
+        return FALSE;
+    }
```
Do this check before getting the properties

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+    HashMap<String, std::variant<bool, unsigned, String>> values;
+    for (unsigned i = 0; i < totalProperties; i++) {
+        GUniqueOutPtr<GError> lookupError;
+        auto propertyName = makeString(properties.get()[i]->name);
```
```suggestion
        auto propertyName = String::fromUTF8(properties.get()[i]->name);
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+    for (unsigned i = 0; i < totalProperties; i++) {
+        GUniqueOutPtr<GError> lookupError;
+        auto propertyName = makeString(properties.get()[i]->name);
+        if (!g_key_file_has_key(keyFile, groupName, propertyName.utf8().data(), &lookupError.outPtr())) {
```
Here we can use properties.get()[i]->name to avoid converting from utf8 and again back to utf8 here.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+
+        switch (G_PARAM_SPEC_VALUE_TYPE(properties.get()[i])) {
+        case G_TYPE_BOOLEAN: {
+            bool value = g_key_file_get_boolean(keyFile, groupName, propertyName.utf8().data(), &lookupError.outPtr());
```
And here again, why have we converted to String?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+                g_propagate_error(error, lookupError.release());
+                return FALSE;
+            }
+            values.add(WTFMove(propertyName), value);
```
Here you can do `values.add(String::fromUTF8(properties.get()[i]->name), value);` and avoid all previous conversions.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+        return FALSE;
+    }
+
+    HashMap<String, std::variant<bool, unsigned, String>> values;
```
```suggestion
    HashMap<String, std::variant<bool, unsigned, CString>> values;
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+                g_propagate_error(error, lookupError.release());
+                return FALSE;
+            }
+            values.add(WTFMove(propertyName), makeString(value.get()));
```
I suggested to use CString instead of string for the value, but maybe we can even use `GUniquePtr<char>` and we don't copy anything?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+    }
+
+    for (unsigned i = 0; i < length; i++) {
+        auto key = makeString(allKeys.get()[i]);
```
```suggestion
        auto key = String::fromUTF8(allKeys.get()[i]);
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+    for (unsigned i = 0; i < length; i++) {
+        auto key = makeString(allKeys.get()[i]);
+        if (!values.contains(key)) {
+            g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_INVALID_VALUE, "The %s group contains an invalid setting: %s", groupName, key.utf8().data());
```
```suggestion
            g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_INVALID_VALUE, "The %s group contains an invalid setting: %s", groupName, allKeys.get()[i]);
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,118 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+            g_object_set(settings, key.utf8().data(), value.utf8().data(), nullptr);
+        });
+    }
+    g_object_thaw_notify(G_OBJECT(settings));
```
Using GValue instead of std::variant we could use g_object_setv

philn commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,119 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+        }
+        }
+        g_ptr_array_add(propertyNames.get(), const_cast<char*>(name));
+        g_array_append_val(values.get(), value);
```
This needs to be done only if the value is set.

philn commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,119 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+                return FALSE;
+            }
+            g_value_init(&value, G_TYPE_STRING);
+            g_value_set_string(&value, stringValue.get());
```
Maybe try `g_value_take_string()`. Updating the patch.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4131,3 +4131,127 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
+
+    g_object_freeze_notify(G_OBJECT(settings));
+    g_object_setv(G_OBJECT(settings), propertyNames->len, (const char**)propertyNames->pdata, reinterpret_cast<GValue*>(values->data));
+    g_object_thaw_notify(G_OBJECT(settings));
```
`g_object_setv` already does the freeze/thaw, so we don't need to do it explicitly.

carlosgcampos replied:
Use C++ style cast too.

philn replied:
Couldn't make C++ casts work :(

carlosgcampos reviewed (approved):
LGTM, please check my comments before landing

philn committed:
[GLib] Support for loading websettings from a config file
https://bugs.webkit.org/show_bug.cgi?id=273043

Reviewed by Michael Catanzaro, Carlos Garcia Campos and Adrian Perez de Castro.

New WebKitSettings API allowing to apply settings declared in a .ini style file. Based on a similar
feature implemented in Cog by Adrian Perez de Castro <aperez@igalia.com>.

* Source/WTF/wtf/glib/GUniquePtr.h:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webkit_settings_apply_from_key_file):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Tools/MiniBrowser/gtk/main.c:
(main):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitSettingsApplyFromConfigFile):
(beforeAll):

Canonical link: https://commits.webkit.org/278688@main
